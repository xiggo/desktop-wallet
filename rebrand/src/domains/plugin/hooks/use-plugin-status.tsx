import { useTranslation } from "react-i18next";

interface Properties {
	isEnabled?: boolean;
	isInstalled?: boolean;
}

const usePluginStatus = () => {
	const { t } = useTranslation();

	const getPluginStatus = ({ isEnabled, isInstalled }: Properties) => {
		if (!isInstalled) {
			return t("PLUGINS.STATUS.NOT_INSTALLED");
		}

		if (isEnabled) {
			return t("PLUGINS.STATUS.ENABLED");
		}

		return t("PLUGINS.STATUS.DISABLED");
	};

	return { getPluginStatus };
};

export { usePluginStatus };
