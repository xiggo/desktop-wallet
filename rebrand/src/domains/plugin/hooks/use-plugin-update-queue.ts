/* eslint-disable arrow-body-style */
import { Contracts } from "@payvo/sdk-profiles";
import { useCallback, useEffect, useRef, useState } from "react";

import { usePluginManagerContext } from "@/plugins/context/PluginManagerProvider";

export const usePluginUpdateQueue = (profile: Contracts.IProfile) => {
	const [queue, setQueue] = useState<any[]>([]);
	const [isUpdating, setIsUpdating] = useState(false);
	const [isUpdateCompleted, setIsUpdateCompleted] = useState(false);
	const { updatePlugin } = usePluginManagerContext();
	const initialQueueReference = useRef<string[]>([]);

	const currentPlugin = queue[0];

	const profileId = profile.id();

	const startUpdate = (plugins: any[]) => {
		const ids = plugins.map((plugin) => plugin.id);
		initialQueueReference.current = ids;

		setQueue(plugins);
		setIsUpdating(true);
		setIsUpdateCompleted(false);
	};

	const next = () => {
		setQueue((current) => current.slice(1));
	};

	const update = useCallback(async () => {
		await updatePlugin(currentPlugin, profileId);
		setTimeout(() => next(), 0);
	}, [currentPlugin, profileId]); // eslint-disable-line react-hooks/exhaustive-deps

	const hasUpdateComplete = (id: string) => {
		return initialQueueReference.current.includes(id) && !hasInUpdateQueue(id);
	};

	const hasInUpdateQueue = (id: string) => {
		return queue.some((plugin) => plugin.id === id);
	};

	useEffect(() => {
		if (!isUpdating) {
			return;
		}

		if (queue.length === 0) {
			setIsUpdating(false);
			setIsUpdateCompleted(true);
			return;
		}

		update();
	}, [isUpdating, queue, update]);

	return {
		hasInUpdateQueue,
		hasUpdateComplete,
		isUpdateCompleted,
		isUpdating,
		startUpdate,
	};
};
