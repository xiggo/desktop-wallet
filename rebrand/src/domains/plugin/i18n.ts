export const translations = {
	CATEGORIES: {
		GAME: "Game",
		GAMING: "Gaming",
		LANGUAGE: "Language",
		OTHER: "Other",
		UTILITY: "Utility",
	},

	DEVELOPER_RESPONSE: "Developer response",

	DISABLE_SUCCESS: `The plugin "{{name}}" has been disabled`,
	ENABLE_FAILURE: `Failed to enable plugin "{{name}}". Reason: {{msg}}.`,
	ENABLE_SUCCESS: `The plugin "{{name}}" has been enabled`,

	FILTERS: {
		CATEGORIES: {
			GAME: "Game",
			OTHER: "Other",
			UTILITY: "Utility",
		},
	},
	MANUAL_INSTALLATION_DISCLAIMER: {
		DISCLAIMER:
			"By accepting the following terms, you hereby assume the risk associated with downloading files and installing said files from a direct URL link. The aforementioned links have neither been tested nor approved by Solar.org.\n\nWe make no warranties, expressed or implied, as to the sustainability, availability, security, of these URLs. We are not responsible for any consequences which may occur from downloading, viewing, or installing files directly from a URL. Solar.org shall bear no liability for any loss suffered by users as a result of clicking an unverified direct URL.\n\nYou acknowledge that clicking unverified URLs may result in you downloading, viewing, or installing content that may have bugs, glitches, lack of functionality, or can cause damage to your machine or result in the loss of data.",
		TITLE: "Disclaimer",
	},
	MINIMUM_VERSION_NOT_SATISFIED:
		"In order to update this plugin please update the Solar Wallet to v{{minimumVersion}}+",

	MODAL_ENABLE_PLUGIN: {
		DESCRIPTION: "Would you like to enable the Plugin?",
		INSTALLATION: "Installation",
	},

	MODAL_INSTALL_PLUGIN: {
		ALLOW: "Allow",
		DESCRIPTION: "This plugin needs the following permissions:",
		DOWNLOAD_FAILURE: `Failed to download plugin "{{name}}"`,
		INSTALL_FAILURE: `Failed to install plugin "{{name}}". Reason: {{msg}}.`,
	},

	MODAL_MANUAL_INSTALL_PLUGIN: {
		DESCRIPTION: "Fetch the Plugin directly from GitHub by providing the repository URL below.",
		ERROR: "Failed to find a valid plugin repository. Please verify the URL and try again.",
		REPOSITORY_URL: "GitHub Repository URL",
		TITLE: "Install from URL",
	},

	MODAL_UNINSTALL: {
		DESCRIPTION:
			"Are you sure you want to remove '{{ name }}'? Uninstalling this plugin will remove it from all profiles on this system.",
		TITLE: "Uninstall Plugin",
	},

	MODAL_UPDATES_CONFIRMATION: {
		DESCRIPTION_COMPATIBLE: "The following plugins will be updated.",
		DESCRIPTION_INCOMPATIBLE:
			"Some of the plugins cannot be updated as they require a newer version of the Solar Wallet. Pressing Continue will install all compatible updates.",
		TITLE: "Plugin Updates",
	},

	NEW_UPDATES_AVAILABLE: "New updates available",

	NEW_VERSION_AVAILABLE: "New version available",

	PAGE_PLUGIN_MANAGER: {
		DESCRIPTION: "An easy way to find, install, and manage Plugins.",
		NO_PLUGINS_AVAILABLE: "This category has no available Plugins yet.",

		NO_PLUGINS_FOUND: `Your search query <strong>{{query}}</strong> does not match any Plugins in the manager.`,
		NO_PLUGINS_INSTALLED: "You haven't installed any Plugins yet. Once you install them, they will show up here.",
		TITLE: "Plugin Manager",

		VIEW: {
			ALL: "All",
			GAMING: "Gaming",
			LATEST: "Latest",
			MY_PLUGINS: "My Plugins",
			OTHER: "Other",
			SEARCH: "Search Results",
			UTILITY: "Utility",
		},
	},

	PERMISSIONS: {
		// Legacy permissions
		ALERTS: "Allows access to the Solar Wallet alerts",
		AUDIO: "Allows access to play audio from within the Solar Wallet",
		AVATARS: "Plugin contains custom avatars",
		COMPONENTS: "Allows loading custom components",
		DIALOGS: "Allows using file dialogs",
		EVENTS: "Allows access to the Solar Wallet events",
		FILESYSTEM: "Allows using file dialogs",
		HTTP: "Allows performing external web requests",
		LANGUAGES: "Allows loading additional languages for the Solar Wallet",
		LAUNCH: "Allows to register a custom view",
		MENU_ITEMS: "Allows adding custom menu items to the Solar Wallet sidebar",
		MESSAGING: "Allows WebFrames access to a one-way messaging system",
		PEER_ALL: "Allows access to the peer discovery",
		PEER_CURRENT: "Allows access to the currently connected peer",
		PROFILE: "Allows access to the currently active profile",
		PROFILE_ALL: "Allows access to all available profiles",
		PROFILE_CURRENT: "Allows access to the currently active profile",
		PUBLIC: "Allows navigation to wallet routes and provides access to the Font Awesome icon set",
		ROUTES: "Allows loading additional routes into the Solar Wallet",
		STORAGE: "Allows storing data within the Solar Wallet, using key-value pairs",
		STORE: "Allows storing data within the Solar Wallet, using key-value pairs",
		THEME: "Allows loading additional custom themes for the Solar Wallet",
		THEMES: "Allows loading additional custom themes for the Solar Wallet",
		TIMERS: "Allows using timeouts and intervals",
		UI_COMPONENTS: "Allows access to the standard Solar Wallet components used throughout",
		UTILS: "Allows using utilities such as the BigNumber type and dayjs",
		WALLET_TABS: "Allows showing an additional tab/page on the Wallet screen",
		WEBFRAME: "Allows showing remote URL pages within a frame",
		WEBSOCKET: "Allows connections to websockets",
	},

	PLUGIN_INFO: {
		ABOUT: "About",
		DISCLAIMER:
			"The availability of this Plugin in the Solar Wallet does not mean that Solar.org is directly involved in its development or developers. By installing it, you assume any associated risks.",
		NOT_ENABLED: "This Plugin is not enabled",
		NOT_LAUNCHEABLE: "This Plugin cannot be launched",
		REPORT: "Report Plugin",
		REQUIREMENTS: "Requirements",
		SCREENSHOTS: "Screenshots",
		WALLET_VERSION: "Solar Wallet Version v{{minimumVersion}}+",
	},

	REQUIRED_VERSION: "Required Version",

	STATUS: {
		COMPATIBLE: "Compatible",
		DISABLED: "Disabled",
		ENABLED: "Enabled",
		INCOMPATIBLE: "Incompatible",
		NOT_INSTALLED: "Not installed",
	},

	SUCCESSFULLY_INSTALLED: "Successfully Installed",

	UPDATE_ALL: "Update All",

	UPDATE_ALL_NOTICE_INCOMPATIBLE_one:
		"There is {{count}} update available for your plugins, but your wallet version is incompatible. Please update the Solar Wallet to continue.",
	UPDATE_ALL_NOTICE_INCOMPATIBLE_other:
		"There are {{count}} updates available for your plugins, but your wallet version is incompatible. Please update the Solar Wallet to continue.",
	UPDATE_ALL_NOTICE_one: "There is {{count}} update available for your plugins. Would you like to update it?",
	UPDATE_ALL_NOTICE_other:
		"There are {{count}} updates available for your plugins. Would you like to update them all now?",

	WARNING_DISCLAIMER:
		"Please make sure to check the documentation of the Plugin before installing it. By installing the Plugin, you assume any associated risks.",
};
