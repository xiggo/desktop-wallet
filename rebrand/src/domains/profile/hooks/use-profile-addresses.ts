import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";
import { useMemo } from "react";

export interface AddressProperties {
	id: string;
	address: string;
	alias?: string;
	network?: string;
	avatar: string;
	type: string;
}

interface ProfileAddressesProperties {
	profile: Contracts.IProfile;
	network?: Networks.Network;
}

const isMultiSignature = (wallet: Contracts.IReadWriteWallet) => {
	try {
		return wallet.isMultiSignature();
	} catch {
		/* istanbul ignore next */
		return false;
	}
};

export const useProfileAddresses = (
	{ profile, network }: ProfileAddressesProperties,
	exceptMultiSignature?: boolean,
) => {
	const contacts = profile.contacts().values();
	const profileWallets = profile.wallets().values();

	return useMemo(() => {
		const allAddresses: AddressProperties[] = [];
		const profileAddresses: AddressProperties[] = [];
		const contactAddresses: AddressProperties[] = [];

		const isNetworkSelected = (addressNetwork: string) => {
			if (!network) {
				return true;
			}

			return addressNetwork === network.id();
		};

		for (const wallet of profileWallets) {
			if (!isNetworkSelected(wallet.network().id())) {
				continue;
			}

			if (exceptMultiSignature && isMultiSignature(wallet)) {
				continue;
			}

			const address = {
				address: wallet.address(),
				alias: wallet.alias(),
				avatar: wallet.avatar(),
				id: wallet.id(),
				network: wallet.network().id(),
				type: "wallet",
			};

			allAddresses.push(address);
			profileAddresses.push(address);
		}

		for (const contact of contacts) {
			for (const contactAddress of contact.addresses().values()) {
				if (!isNetworkSelected(contactAddress.network())) {
					continue;
				}

				if (exceptMultiSignature) {
					continue;
				}

				const addressAlreadyExist = allAddresses.some(({ address }) => address === contactAddress.address());
				if (addressAlreadyExist) {
					continue;
				}

				const address = {
					address: contactAddress.address(),
					alias: contact.name(),
					avatar: contactAddress.avatar(),
					id: contactAddress.id(),
					network: contactAddress.network(),
					type: "contact",
				};

				allAddresses.push(address);
				contactAddresses.push(address);
			}
		}

		return {
			allAddresses,
			contactAddresses,
			profileAddresses,
		};
	}, [network, profileWallets, exceptMultiSignature, contacts]);
};
