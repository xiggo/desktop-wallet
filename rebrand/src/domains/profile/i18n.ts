export const translations = {
	CREATE_PROFILE: "Create Profile",
	IMPORT: {
		ERROR_STEP: {
			DESCRIPTION:
				"An error occurred while importing the profile, try importing again. Select Retry to re-import the profile. If the error occurs again, you can contact our support team.",
		},
		FORM_STEP: {
			DESCRIPTION: "Your profile addresses have been imported.",
		},

		PASSWORD_DESCRIPTION:
			"The profile you are importing contains a password that has been set for security purposes. Enter the password to confirm.",
		PASSWORD_TITLE: "Profile Password",
		PROCESSING_IMPORT_STEP: {
			DESCRIPTION: "The file with name {{name}} is being imported.",
			PLEASE_WAIT: "Please wait...",
		},
		SELECT_FILE_STEP: {
			BROWSE_FILES: "Browse Files",
			CLICK_HERE: "Click here.",
			DEPRECATION_WARNING:
				"Please note that importing a Profile from Desktop Wallet v2 in .json format is supported temporarily and will soon be disabled. We recommend that you re-save your Profile in the new .dwe format.",
			DESCRIPTION:
				"Select a {{fileFormat}} file with your Profile and related properties to start the import process.",
			DRAG_AND_DROP: "Drag & Drop or",
			ERRORS: {
				NOT_SUPPORTED:
					"This file type is not supported.\nPlease retry with a <strong>{{fileFormat}}</strong> file.",
				TOO_MANY: "You tried to import <strong>{{fileCount}}</strong> files.\nPlease retry with one file only.",
			},
			LEGACY_IMPORT: "Importing from Desktop Wallet v2?",
			SUPPORTED_FORMAT: "Supported format is {{fileFormat}}",
			UPLOAD_TITLE: "Click here to upload",
		},
		TITLE: "Import Profile",
	},
	MODAL_DELETE_PROFILE: {
		DESCRIPTION: "Do you really want to delete this profile? Once deleted, you will not be able to restore it.",
		TITLE: "Delete Profile",
	},
	MODAL_HISTORY: {
		TITLE: "History",
		TYPES: {
			REGISTRATION: "Registration",
			RESIGN: "Resign",
			UPDATE: "Update",
		},
	},
	MODAL_RESET_PROFILE: {
		DESCRIPTION:
			"By performing this action, all of your Profile settings will be restored to default. This won't delete your wallets or plugins. This action cannot be undone. Do you want to reset your settings?",
		SUCCESS: "Your profile has been reset successfully",
		TITLE: "Reset Profile Settings",
	},
	MODAL_SELECT_SENDER: {
		DESCRIPTION: "Find and select the sender from your wallets",
		SEARCH_PLACEHOLDER: "Enter name or address",
		TITLE: "Select Sender",
	},
	MODAL_SIGN_IN: {
		DESCRIPTION: "Sign in to access your Profile.",
		MAX_ATTEMPTS_ERROR:
			"Maximum sign in attempts reached - please wait {{remainingTime}} seconds before trying again.",
		TITLE: "Sign In",
	},
	MODAL_WELCOME: {
		DONT_SHOW_CHECKBOX_LABEL: "Don't show this quick tour again.",
		STEP_1_DESCRIPTION:
			"Start a quick tour to learn how to get the most out of the Solar Wallet, or skip it to go directly to your profile.",
		STEP_1_TITLE: "Welcome to Solar Wallet",
		STEP_2_DESCRIPTION:
			"Create or import existing wallets from a broad range of assets and secure them using industry-standard encryption.",
		STEP_2_TITLE: "Manage Your Assets",
		STEP_3_DESCRIPTION:
			"Solar lets you extend your wallet’s functionality with a growing collection of Games, Utilities and more.",
		STEP_3_TITLE: "Enhance Your Experience With Plugins",
		STEP_4_DESCRIPTION: "With Solar, you’re in control. Swap assets effortlessly without ever leaving your wallet.",
		STEP_4_TITLE: "Exchange Assets On The Fly",
		STEP_5_DESCRIPTION:
			"Customize or search your news feed to view the latest updates on your favorite cryptoassets.",
		STEP_5_TITLE: "Stay Up To Date",
	},
	PAGE_CREATE_PROFILE: {
		DESCRIPTION: "Create a new Profile below.",
		TITLE: "Create Profile",
		VALIDATION: {
			NAME_EXISTS: "A profile with this name already exists",
		},
	},
	PAGE_WELCOME: {
		HAS_EXPORTED_PROFILES: "Exported a Profile before?",
		IMPORT_PROFILE: "Import it here.",
		IMPORT_PROFILE_TITLE: "Import Profile",
		WITHOUT_PROFILES: {
			DESCRIPTION: "Create a new Profile to get started.",
			TITLE: "Create Profile",
		},
		WITH_PROFILES: {
			DESCRIPTION: "Choose from an existing Profile below or create a new Profile to get started.",
			TITLE: "Select Profile",
		},
	},
};
