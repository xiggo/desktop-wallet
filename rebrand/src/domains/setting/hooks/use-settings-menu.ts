import { useTranslation } from "react-i18next";

export const useSettingsMenu = () => {
	const { t } = useTranslation();

	const menuItems = [
		{
			icon: "Sliders",
			itemKey: "general",
			label: t("SETTINGS.GENERAL.MENU_ITEM"),
			route: "general",
		},
		{
			icon: "Lock",
			itemKey: "password",
			label: t("SETTINGS.PASSWORD.MENU_ITEM"),
			route: "password",
		},
		{
			icon: "ArrowUpTurnBracket",
			itemKey: "export",
			label: t("SETTINGS.EXPORT.MENU_ITEM"),
			route: "export",
		},
		{
			icon: "PencilRuler",
			itemKey: "appearance",
			label: t("SETTINGS.APPEARANCE.MENU_ITEM"),
			route: "appearance",
		},
	];

	return { menuItems };
};
