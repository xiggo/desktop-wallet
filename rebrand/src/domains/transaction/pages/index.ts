export { SendDelegateResignation } from "./SendDelegateResignation";
export { SendIpfs } from "./SendIpfs";
export { SendRegistration } from "./SendRegistration";
export { SendTransfer } from "./SendTransfer";
export { SendVote } from "./SendVote";
