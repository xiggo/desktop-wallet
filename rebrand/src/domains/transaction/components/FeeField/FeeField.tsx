import { Networks } from "@payvo/sdk";
import { isEqual } from "@payvo/sdk-helpers";
import { Contracts } from "@payvo/sdk-profiles";
import React, { useEffect, useMemo, useRef, useState } from "react";
import { useFormContext } from "react-hook-form";
import { useTranslation } from "react-i18next";

import { useDebounce, useFees } from "@/app/hooks";
import { toasts } from "@/app/services";
import { InputFee } from "@/domains/transaction/components/InputFee";

interface Properties {
	type: string;
	data: Record<string, any> | undefined;
	network: Networks.Network;
	profile: Contracts.IProfile;
}

export const FeeField: React.FC<Properties> = ({ type, network, profile, ...properties }: Properties) => {
	const isMounted = useRef(true);
	const { t } = useTranslation();

	const { calculate } = useFees(profile);

	const [isLoadingFee, setIsLoadingFee] = useState(false);

	const { setError, clearErrors, watch, setValue, getValues } = useFormContext();
	const { fees, inputFeeSettings = {} } = watch(["fees", "inputFeeSettings"]);

	// getValues does not get the value of `defaultValues` on first render
	const [defaultFee] = useState(() => watch("fee"));
	const fee = getValues("fee") || defaultFee;

	const [data, isLoadingData] = useDebounce(properties.data, 700);

	const isCalculatingFee = useMemo(() => isLoadingData || isLoadingFee, [isLoadingData, isLoadingFee]);
	const usesSizeBasedFee = useMemo(() => network.feeType() === "size", [network]);
	const usesStaticFee = useMemo(() => network.feeType() !== "dynamic" || !fees?.isDynamic, [fees, network]);

	const showFeeChangedToast = () => {
		toasts.warning(t("TRANSACTION.PAGE_TRANSACTION_SEND.FORM_STEP.FEE_UPDATE"));
	};

	useEffect(() => {
		if (usesSizeBasedFee || type === "multiSignature") {
			if (isCalculatingFee) {
				// This message does not need i18n as it is only intended to mark
				// the form as invalid until fee calculation has been completed.
				setError("feeCalculation", { message: "fee calculation not completed" });
			} else {
				clearErrors("feeCalculation");
			}
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [isCalculatingFee, setError, clearErrors, usesSizeBasedFee]);

	useEffect(() => {
		const recalculateFee = async () => {
			// Set fees to 0 when fee type is "size" but current form data
			// is not sufficient yet to calculate the transaction fees.
			if (usesSizeBasedFee) {
				const resetFees = () => {
					setValue("fees", { avg: 0, max: 0, min: 0, static: 0 });
				};

				if (data === undefined) {
					return resetFees();
				}

				if (type === "transfer" && (!data.amount || !data.to)) {
					return resetFees();
				}

				if (type === "multiPayment" && !data.payments?.length) {
					return resetFees();
				}

				if (type === "vote" && !data.votes?.length && !data.unvotes?.length) {
					return resetFees();
				}

				if (type === "delegateRegistration" && !data.username) {
					return resetFees();
				}

				if (type === "secondSignature" && !data.mnemonic) {
					return resetFees();
				}
			}

			setIsLoadingFee(true);

			const transactionFees = await calculate({
				coin: network.coin(),
				data,
				network: network.id(),
				type,
			});

			if (!isEqual(getValues("fees"), transactionFees)) {
				if (
					network.feeType() === "static" ||
					transactionFees.isDynamic === false ||
					getValues("fee") === undefined
				) {
					const newFee = transactionFees.isDynamic ? transactionFees.avg : transactionFees.static;

					if (getValues("fee") !== undefined) {
						showFeeChangedToast();
					}

					setValue("fee", newFee, { shouldDirty: true, shouldValidate: true });
				}

				if (usesSizeBasedFee && +getValues("fee") < +transactionFees.min) {
					setValue("fee", transactionFees.min, { shouldDirty: true, shouldValidate: true });

					showFeeChangedToast();
				}

				setValue("fees", transactionFees, { shouldDirty: true, shouldValidate: true });
			}

			/* istanbul ignore next */
			if (isMounted.current) {
				setIsLoadingFee(false);
			}
		};

		recalculateFee();
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [calculate, data, getValues, isMounted, network, setValue, type, usesSizeBasedFee]);

	useEffect(
		/* istanbul ignore next */
		() => () => {
			isMounted.current = false;
		},
		[],
	);

	return (
		<InputFee
			min={fees?.min}
			avg={fees?.avg}
			max={fees?.max}
			loading={!fees || isLoadingFee}
			value={fee}
			step={0.01}
			disabled={usesStaticFee && !usesSizeBasedFee}
			network={network}
			profile={profile}
			onChange={(value) => {
				setValue("fee", value, { shouldDirty: true, shouldValidate: true });
			}}
			viewType={inputFeeSettings.viewType}
			onChangeViewType={(viewType) => {
				setValue(
					"inputFeeSettings",
					{ ...inputFeeSettings, viewType },
					{ shouldDirty: true, shouldValidate: true },
				);
			}}
			simpleValue={inputFeeSettings.simpleValue}
			onChangeSimpleValue={(simpleValue) => {
				setValue(
					"inputFeeSettings",
					{ ...inputFeeSettings, simpleValue },
					{ shouldDirty: true, shouldValidate: true },
				);
			}}
		/>
	);
};
