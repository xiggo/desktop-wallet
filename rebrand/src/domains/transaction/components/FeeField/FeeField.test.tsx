import { Networks } from "@payvo/sdk";
import { ARK } from "@payvo/sdk-ark";
import { LSK } from "@payvo/sdk-lsk";
import { Contracts } from "@payvo/sdk-profiles";
import userEvent from "@testing-library/user-event";
import React from "react";
import { FormProvider, useForm } from "react-hook-form";

import { useValidation } from "@/app/hooks";
import * as useFeesHook from "@/app/hooks/use-fees";
import { toasts } from "@/app/services";
import { FeeField } from "@/domains/transaction/components/FeeField/FeeField";
import { translations } from "@/domains/transaction/i18n";
import { env, getDefaultProfileId, render, screen, waitFor } from "@/utils/testing-library";

describe("FeeField", () => {
	let profile: Contracts.IProfile;

	const networksByFeeType = {
		default: new Networks.Network(ARK.manifest, ARK.manifest.networks["ark.devnet"]),
		size: new Networks.Network(LSK.manifest, LSK.manifest.networks["lsk.testnet"]),
	};

	beforeAll(() => {
		env.registerCoin("LSK", LSK);
		profile = env.profiles().findById(getDefaultProfileId());
	});

	const Component = ({ balance = 10, network = networksByFeeType.default, type, data }: any) => {
		const form = useForm({ mode: "onChange" });

		const { register, watch } = form;
		const { common } = useValidation();
		const { fees } = watch();

		register("fee");
		register("fees", common.fee(balance, network, fees));
		register("inputFeeSettings");

		return (
			<FormProvider {...form}>
				<FeeField type={type} data={data} network={network} profile={profile} />
			</FormProvider>
		);
	};

	it("should render", async () => {
		const { asFragment } = render(<Component type="transfer" />);

		await waitFor(() => expect(screen.getAllByTestId("Amount")).toHaveLength(3));

		expect(asFragment()).toMatchSnapshot();
	});

	describe("when network's fee type is size", () => {
		it.each(["transfer", "multiPayment", "vote", "delegateRegistration", "secondSignature"])(
			"should show 0 when %s data is undefined",
			async (transactionType) => {
				render(<Component type={transactionType} network={networksByFeeType.size} data={undefined} />);

				await waitFor(() => expect(screen.getAllByTestId("Amount")).toHaveLength(3));

				expect(screen.getAllByTestId("Amount")[0]).toHaveTextContent("0 LSK");
				expect(screen.getAllByTestId("Amount")[1]).toHaveTextContent("0 LSK");
				expect(screen.getAllByTestId("Amount")[2]).toHaveTextContent("0 LSK");
			},
		);

		it.each(["transfer", "multiPayment", "vote", "delegateRegistration", "secondSignature"])(
			"should show 0 %s data is not available yet",
			async (transactionType) => {
				render(<Component type={transactionType} network={networksByFeeType.size} data={{}} />);

				await waitFor(() => expect(screen.getAllByTestId("Amount")).toHaveLength(3));

				expect(screen.getAllByTestId("Amount")[0]).toHaveTextContent("0 LSK");
				expect(screen.getAllByTestId("Amount")[1]).toHaveTextContent("0 LSK");
				expect(screen.getAllByTestId("Amount")[2]).toHaveTextContent("0 LSK");
			},
		);

		it("should recalculate fees on data changes", async () => {
			const calculate = jest.fn().mockResolvedValue({ avg: 2, max: 2, min: 2, static: 2 });
			const useFeesMock = jest.spyOn(useFeesHook, "useFees").mockImplementation(() => ({ calculate }));

			const properties = { network: networksByFeeType.size, type: "transfer" };

			const { rerender } = render(<Component {...properties} data={{}} />);

			await waitFor(() => expect(screen.getAllByTestId("Amount")[0]).toHaveTextContent("0 LSK"));

			rerender(
				<Component {...properties} data={{ amount: 1, to: "lsktpbzum9d9gnhqu3homwbwo8h238zeo3bhpjocg" }} />,
			);

			await waitFor(() => expect(screen.getAllByTestId("Amount")[0]).toHaveTextContent("2 LSK"));

			calculate.mockRestore();
			useFeesMock.mockRestore();
		});

		it("should override fee when it is lower than the minimum fees", async () => {
			let useFeesSpy: jest.SpyInstance;

			const toastSpy = jest.spyOn(toasts, "warning").mockImplementation();

			useFeesSpy = jest.spyOn(useFeesHook, "useFees").mockReturnValue({
				calculate: () => Promise.resolve({ avg: 3, max: 5, min: 1, static: 3 }),
			});

			const { rerender } = render(
				<Component
					network={networksByFeeType.size}
					type="transfer"
					data={{ amount: 1, to: "lsktpbzum9d9gnhqu3homwbwo8h238zeo3bhpjocg" }}
				/>,
			);

			await waitFor(() => expect(screen.getAllByTestId("Amount")[0]).toHaveTextContent("1 LSK"));

			expect(screen.getAllByTestId("Amount")[1]).toHaveTextContent("3 LSK");
			expect(screen.getAllByTestId("Amount")[2]).toHaveTextContent("5 LSK");
			expect(screen.getAllByRole("radio")[1]).toBeChecked();

			userEvent.click(screen.getByText(translations.INPUT_FEE_VIEW_TYPE.ADVANCED));

			await waitFor(() => expect(screen.getByTestId("InputCurrency")).toHaveValue("3"));

			useFeesSpy = jest.spyOn(useFeesHook, "useFees").mockReturnValue({
				calculate: () => Promise.resolve({ avg: 9, max: 12, min: 6, static: 9 }),
			});

			rerender(
				<Component
					network={networksByFeeType.size}
					type="transfer"
					data={{ amount: 2, to: "lsktpbzum9d9gnhqu3homwbwo8h238zeo3bhpjocg" }}
				/>,
			);

			await waitFor(() => expect(screen.getByTestId("InputCurrency")).toHaveValue("6"));

			expect(toastSpy).toHaveBeenCalledWith(translations.PAGE_TRANSACTION_SEND.FORM_STEP.FEE_UPDATE);

			toastSpy.mockRestore();
			useFeesSpy.mockRestore();
		});
	});

	it("should set fee to fees.avg when it has no value yet", async () => {
		const calculate = jest.fn().mockResolvedValue({ avg: 30, max: 1, min: 1, static: 1 });
		const useFeesMock = jest.spyOn(useFeesHook, "useFees").mockImplementation(() => ({ calculate }));

		render(
			<Component
				network={networksByFeeType.size}
				type="transfer"
				data={{ amount: 1, to: "lsktpbzum9d9gnhqu3homwbwo8h238zeo3bhpjocg" }}
			/>,
		);

		await waitFor(() => expect(screen.getAllByTestId("Amount")[0]).toHaveTextContent("1 LSK"));

		expect(screen.getByRole("radio", { checked: true })).toHaveTextContent("30 LSK");

		calculate.mockRestore();
		useFeesMock.mockRestore();
	});
});
