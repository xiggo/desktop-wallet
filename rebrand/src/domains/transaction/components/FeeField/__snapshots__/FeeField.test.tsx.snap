// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeField should render 1`] = `
<DocumentFragment>
  <div
    class="relative"
    data-testid="InputFee"
  >
    <div
      class="absolute right-0 -mt-7"
    >
      <div
        class="flex items-center"
      >
        <span
          class="Switch__SwitchText-sc-161iiro-0 aJKSk"
          role="button"
        >
          Simple
        </span>
        <label
          class="Toggle__Wrapper-sc-6s3wfu-0 iggSjt"
        >
          <input
            class="Toggle__Input-sc-6s3wfu-1 WwplI"
            type="checkbox"
          />
          <div
            class="Toggle__Handle-sc-6s3wfu-2 jRjdQR"
          >
            <span
              class="Toggle__HandleInner-sc-6s3wfu-3 fiPeMI"
            />
          </div>
        </label>
        <span
          class="Switch__SwitchText-sc-161iiro-0 jeOrUV"
          role="button"
        >
          Advanced
        </span>
      </div>
    </div>
    <div
      class="inline-flex items-center space-x-3 w-full"
      data-testid="ButtonGroup"
      role="radiogroup"
    >
      <button
        aria-checked="false"
        aria-label="0.00357"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-2f0qmd-0 ZQFGq"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        <div
          class="flex flex-col space-y-2"
        >
          <div>
            Slow
          </div>
          <span
            class="whitespace-nowrap text-sm"
            data-testid="Amount"
          >
            0.00357 DARK
          </span>
        </div>
      </button>
      <button
        aria-checked="true"
        aria-label="0.07320598"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-2f0qmd-0 ZQFGq"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        <div
          class="flex flex-col space-y-2"
        >
          <div>
            Average
          </div>
          <span
            class="whitespace-nowrap text-sm"
            data-testid="Amount"
          >
            0.07320598 DARK
          </span>
        </div>
      </button>
      <button
        aria-checked="false"
        aria-label="0.1"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-2f0qmd-0 ZQFGq"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        <div
          class="flex flex-col space-y-2"
        >
          <div>
            Fast
          </div>
          <span
            class="whitespace-nowrap text-sm"
            data-testid="Amount"
          >
            0.1 DARK
          </span>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
